<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="test_single_motor">

  <link name="world">
    </link>

    <link name="box">
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
            <geometry>
                <box size="0.5 0.5 0.5"/>
            </geometry>
            <material name="red">
                <color rgba="1.0 0.0 0.0 1.0"/>
            </material>
        </visual>
    </link>

    <joint name="world_box_1" type="fixed">
        <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
        <parent link="world"/>
        <child link="box"/>
    </joint>

    <link name="rotor">
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
            <geometry>
                <cylinder radius="0.2" length="0.1"/>
            </geometry>
            <material name="red">
                <color rgba="0.0 0.0 1.0 1.0"/>
            </material>
        </visual>
    </link>

    <link name="indicator">
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
            <geometry>
                <box size="0.01 0.3 0.01"/>
            </geometry>
            <material name="">
                <color rgba="0.0 1.0 0.0 1.0"/>
                <texture filename=""/>
            </material>
        </visual>
    </link>

    <joint name="joint_1" type="revolute">
        <origin xyz="0.0 0.3 0.0" rpy="-1.5707 0.0 0.0"/>
        <parent link="box"/>
        <child link="rotor"/>
        <axis xyz="0.0 0.0 -1.0"/>
        <limit lower="-10.0" upper="10.0" effort="10.0" velocity="100.0"/>
    </joint>

    <joint name="indicator_joint" type="fixed">
        <origin xyz="0.0 -0.25 0.055" rpy="0.0 0.0 0.0"/>
        <parent link="rotor"/>
        <child link="indicator"/>
    </joint>


    <link name="box_2">
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
            <geometry>
                <box size="0.5 0.5 0.5"/>
            </geometry>
            <material name="red">
                <color rgba="1.0 0.0 0.0 1.0"/>
            </material>
        </visual>
    </link>

    <joint name="world_box_2" type="fixed">
        <origin xyz="1.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
        <parent link="world"/>
        <child link="box_2"/>
    </joint>

    <link name="rotor_2">
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
            <geometry>
                <cylinder radius="0.2" length="0.1"/>
            </geometry>
            <material name="red">
                <color rgba="0.0 0.0 1.0 1.0"/>
            </material>
        </visual>
    </link>

    <link name="indicator_2">
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
            <geometry>
                <box size="0.01 0.3 0.01"/>
            </geometry>
            <material name="">
                <color rgba="0.0 1.0 0.0 1.0"/>
                <texture filename=""/>
            </material>
        </visual>
    </link>

    <joint name="joint_2" type="revolute">
        <origin xyz="0.0 0.3 0.0" rpy="-1.5707 0.0 0.0"/>
        <parent link="box_2"/>
        <child link="rotor_2"/>
        <axis xyz="0.0 0.0 -1.0"/>
        <limit lower="-10.0" upper="10.0" effort="10.0" velocity="100.0"/>
    </joint>

    <joint name="indicator_joint_2" type="fixed">
        <origin xyz="0.0 -0.25 0.055" rpy="0.0 0.0 0.0"/>
        <parent link="rotor_2"/>
        <child link="indicator_2"/>
    </joint>

  <ros2_control name="pi3hat_hardware_interface" type="system">
    <hardware>
          <plugin>pi3hat_hardware_interface/Pi3HatHardwareInterface</plugin>
          <param name="imu_mounting_deg.yaw">0</param>
          <param name="imu_mounting_deg.pitch">0</param>
          <param name="imu_mounting_deg.roll">0</param>
      </hardware>

      <joint name="joint_1">
          <param name="controller_type">moteus</param>
          <param name="controller_can_bus">1</param>
          <param name="controller_can_id">1</param>


          <param name="motor_direction">1</param>
          <param name="motor_position_offset">0.0</param>
          <param name="motor_position_max">500.0</param>
          <param name="motor_position_min">-500.0</param>
          <param name="motor_velocity_max">10.0</param>
          <param name="motor_torque_max">1.0</param>


          <command_interface name="position"/>
          <command_interface name="velocity"/>
          <command_interface name="effort"/>

          <state_interface name = "position"/>
          <state_interface name = "velocity"/>
          <state_interface name = "effort"/>
          <state_interface name = "temperature"/>
      </joint>

      <joint name="joint_2">
          <param name="controller_type">moteus</param>
          <param name="controller_can_bus">2</param>
          <param name="controller_can_id">2</param>


          <param name="motor_direction">1</param>
          <param name="motor_position_offset">0.0</param>
          <param name="motor_position_max">500.0</param>
          <param name="motor_position_min">-500.0</param>
          <param name="motor_velocity_max">10.0</param>
          <param name="motor_torque_max">1</param>


          <command_interface name="position"/>
          <command_interface name="velocity"/>
          <command_interface name="effort"/>

          <state_interface name = "position"/>
          <state_interface name = "velocity"/>
          <state_interface name = "effort"/>
          <state_interface name = "temperature"/>
      </joint>
      <transmission name="fbl_transmission">
          <plugin>transmission_interface/FourBarLinkageTransmission</plugin>
          <actuator name="joint_1_motor" role="actuator1">
            <mechanical_reduction>16.0</mechanical_reduction>
          </actuator>
          <actuator name="joint_2_motor" role="actuator2">
            <mechanical_reduction>16.0</mechanical_reduction>
          </actuator>
          <joint name="joint_1" role="joint1">
            <offset>0.0</offset>
            <mechanical_reduction>1.0</mechanical_reduction>
          </joint>
          <joint name="joint_2" role="joint2">
            <offset>0.0</offset>
            <mechanical_reduction>1.0</mechanical_reduction>
          </joint>
        </transmission>

      <sensor name="imu_sensor">
          <state_interface name = "orientation.x"/>
          <state_interface name = "orientation.y"/>
          <state_interface name = "orientation.z"/>
          <state_interface name = "orientation.w"/>
          <state_interface name = "angular_velocity.x"/>
          <state_interface name = "angular_velocity.y"/>
          <state_interface name = "angular_velocity.z"/>
          <state_interface name = "linear_acceleration.x"/>
          <state_interface name = "linear_acceleration.y"/>
          <state_interface name = "linear_acceleration.z"/>
      </sensor>
  </ros2_control>

</robot>